### PR TITLE
1.2 improvements

### PR DESCRIPTION
**Pull request by ecino**
Few issues corrected and improvements including:

- When changing an active contract (payment_term or contract_lines), a problem was setting a wrong next_invoice_date and was not syncing well GP.
- Various GP sync problems found and fixed.
- Autocompletion : invoices generated for contact persons instead of companies
- Add ability to link invoice_line to contract manually for more flexibility
- Encoding error of export_tools (lsv,dd)
- Improved CRM views to better track origin of sponsorships (add salesperson, won sponsorships, ...)

### CODE REVIEW

**Comment by CSester** [comment]
Are you going to fix Travis before my review? Or should I review quickly?

**Comment by ecino** [comment]
Travis is fixed.

**Comment by ecino** [comment]
Ok, this review is needed soon because we will need to put in production version 1.4 in three weeks. And by the time both 1.3+1.4 should be reviewed (and they contain lot of changes :o:)

Thank you for your help ;-)

File: account_statement_completion_compassion/model/statement.py [hunk 1]
```diff
@@ -151,16 +155,20 @@ def get_from_amount(self, cr, uid, id, st_line, context=None):
                 [('type', '=', 'in_invoice'), ('state', '=', 'open'),
                  ('amount_total', '=', abs(amount))], context=context)
             res = {}
+            partner_obj = self.pool.get('res.partner')
             if invoice_ids:
                 if len(invoice_ids) == 1:
```
Comment (CSester) [inline account_statement_completion_compassion/model/statement.py]: Test here seems to be useless. You can simply keep the else part. The loop checking that all invoices will have no side effect on the len = 1 case.

File: account_statement_completion_compassion/model/statement.py [hunk 2]
```diff
@@ -259,6 +268,26 @@ def _generate_invoice(self, cr, uid, st_line, partner, context=None):
                     wf_service.trg_validate(
                         uid, 'account.invoice', invoice_id, 'invoice_open',
                         cr)
+                elif product.name == GIFT_TYPES[0]:
```
Comment (CSester) [inline account_statement_completion_compassion/model/statement.py]: Is GIFT_TYPES[0] birthday gift? If yes, please add a comment.

File: account_statement_completion_compassion/model/statement.py [hunk 3]
```diff
@@ -259,6 +268,26 @@ def _generate_invoice(self, cr, uid, st_line, partner, context=None):
...
+                    # Set date of invoice two months before child's birthdate
+                    child_birthdate = res.get('child_birthdate')
+                    if child_birthdate:
+                        inv_date = datetime.strptime(st_line['date'], DF)
+                        birthdate = datetime.strptime(child_birthdate, DF)
+                        new_date = inv_date
```
Comment (CSester) [inline account_statement_completion_compassion/model/statement.py]: I don't get the purpose of all this. If you simply want to substract 2 months from birthdate, why not simple do
```python
inv_date = child_birthdate + relativedelta(months=-2)
```
? relativedelta handles inter-year deltas !

File: partner_compassion/model/partner_compassion.py [hunk 4]
```diff
@@ -18,12 +18,19 @@
 
 
 class ResPartner(orm.Model):
+
```
Comment (CSester) [inline partner_compassion/model/partner_compassion.py]: Why this useless line?

File: partner_compassion/model/partner_compassion.py [hunk 5]
```diff
@@ -18,12 +18,19 @@
...
     """ This class upgrade the partners to match Compassion needs.
         It also synchronize all changes with the MySQL server of GP.
     """
 
     _inherit = 'res.partner'
 
+    def _lang_get(self, cr, uid, context=None):
+        lang_pool = self.pool.get('res.lang')
```
Comment (CSester) [inline partner_compassion/model/partner_compassion.py]:
```python
lang_obj = ...
```
to stay consistent.

**Comment by CSester** [comment]
Good job. :+1:

File: account_statement_completion_compassion/model/statement.py (hunk 3 above)
Comment (ecino) [inline account_statement_completion_compassion/model/statement.py]: In fact it is not as easy as the comment suggest it... if the payment is made less than three months after the birthday, the invoice_date will be at the same date of the payment.
In other situation, it is set two months before the birthday.

Moreover, child_birthdate has a very old year, we should keep the current year for the invoice, or next year if birthday is in future.

Comment (CSester) [inline account_statement_completion_compassion/model/statement.py]: Ok, so if the code fits the needs, please update the comment. A bad comment is worse than no comment at all...

**Comment by CSester** [comment]
Emanuel, you can merge this one. This way, the review for 1.3 and 1.4 will be easier for me.

**Comment by ecino** [comment]
I'll fix the bad comment in a following branch.

**Pull request merged**